nemo/collections/duplex_s2s/modules/speech_generation.py

```diff
@@ -0,0 +1,185 @@
+import torch
+from omegaconf import DictConfig
+from torch import nn
+
+from nemo.core.classes.module import NeuralModule
+from nemo.collections.tts.modules import t5tts_transformer
+
+class TransformerARSpeechDecoder(NeuralModule):
+    def __init__(self, speech_decoder_parms: DictConfig, lantent_dim: int, num_audio_codebooks: int, num_audio_tokens_per_codebook: int):
+        super().__init__()
+        self.use_input_cache = False
+        self.speech_decoder_parms = speech_decoder_parms
+        self.lantent_dim = lantent_dim
+        self.num_audio_codebooks = num_audio_codebooks
+        self.num_audio_tokens_per_codebook = num_audio_tokens_per_codebook
+        # optional configs
+        self.cfg_unconditional_prob = self.speech_decoder_parms.pop("cfg_unconditional_prob", None)
+        self.cfg_scale = self.speech_decoder_parms.pop("cfg_scale", 2.5)
+        self.cond_on_prev_audio_tokens = self.speech_decoder_parms.pop("cond_on_prev_audio_tokens", True)
+        self.detach_input = self.speech_decoder_parms.pop("detach_input", False)
+
+        # projection to adapt llm embeddings into the same shape of speech decoder expected input
+        if lantent_dim != self.speech_decoder_parms["d_model"]:
+            self.input_proj = nn.Linear(lantent_dim, self.speech_decoder_parms["d_model"])
+        else:
+            self.input_proj = None
+
+        # instanciate T5-TTS decoder to full compatibility and potentialy load pretrained model
+        self.t5_decoder = t5tts_transformer.Transformer(**self.speech_decoder_parms)
+
+        # projection to predict audio codes
+        self.final_proj = nn.Linear(self.speech_decoder_parms["d_model"], num_audio_codebooks * num_audio_tokens_per_codebook)
+
+        # create embeddings for encode input tokens
+        if self.cond_on_prev_audio_tokens:
+            audio_embeddings = []
+            for _ in range(self.num_audio_codebooks):
+                audio_embeddings.append(nn.Embedding(num_audio_tokens_per_codebook, self.speech_decoder_parms["d_model"]))
+
+            self.audio_embeddings = nn.ModuleList(audio_embeddings)
+
+    def forward(self, hidden_states, speech_mask, input_audio_tokens=None, return_raw_logits=False):
+        # Megatron LLM parallel training returns T, B, F so reshape it
+        # T, B, F = hidden_states.size()
+        hidden_states = hidden_states.transpose(0, 1).contiguous() # .reshape(B, T, F) # from [T, B, F] to [B, T, F]
+        # input cache needed due our transformer kv cache implementation expect the whole left context
+        if self.use_input_cache:
```

Comment:
Can we transform stateful KV cache to argument+return value for a pure functional API?

Reply:
Deferred for follow up work